### DISCUSSION TITLE
Can I skip XCODE_WORKSPACE / PROJECT?

### DISCUSSION THREAD

**Discussion by morsagmon** [codemagic-ci-cd/codemagic-docs#1279] [Q&A] [answered]
Hello team.
When writing a workflow for iOS, it is mentioned to define vars in the environment:
```
XCODE_WORKSPACE: 
XCODE_SCHEME:
```
However, what if I never used Xcode with my app?
It was developed in VSCode IDE and I want CodeMagic to take the code and build it without me having to upload the app to XCode and manage the project there.
(I only used my Mac to generate the CSR for the certificate and profile keys).

How would my yaml workflow look like in this case, without these vars?
Thanks!

**Comment by icarusdust** [comment, marked as answer]
Hi @morsagmon, each iOS app has its scheme/s and workspace and your project built upon them. You must use Xcode to debug your app locally. You need to make sure that you app runs ok locally and only then you should try it on a CI/CD.

**Reply by morsagmon** [reply]
Thanks!

**Comment by morsagmon** [comment]
Getting back to this issue.
I have manually deployed the app to an ios device using xcode.
Still, I'm not sure what values should I set to these two vars in the codemagic yaml file:

XCODE_WORKSPACE: 
XCODE_SCHEME:

I guess I have a confusion about the way codemagic builds the deployed code. I assumed that it uses its own xcode on its own mac_mini machine, having nothing to do with my (local) xcode. So, are these two vars values coming from my local xcode (doesn't make sense) or something else?

**Reply by icarusdust** [reply]
Xcode is a development toolset, you develop your iOS app in Xcode and store it in Github and then Codemagic references it from Github, your local Xcode does not have any role in a build process of Codemagic. These two values are your local project workspace and scheme, they are not Codemagic specific but general iOS development specific. When building iOS projects locally and running `xcodebuild` command, you are still required to provide `XCODE_WORKSPACE`: `xcodebuild -workspace <your_workspace_name>.xcworkspace`
Please read the following two about Xcode workspace and Schemes:
https://developer.apple.com/library/archive/featuredarticles/XcodeConcepts/Concept-Workspace.html
https://developer.apple.com/library/archive/featuredarticles/XcodeConcepts/Concept-Schemes.html

**Reply by morsagmon** [reply]
Thanks for your reply, icarusdust.
So, if I understand correctly:
- The xcode variables I used on my local machine to build and run the app has nothing to do whatsoever with the variables I define for codemagic in the yaml file.
- The variables I define for codemagic in the yaml file will be used by the mac_mini server when building the app.
- The values for these two variables follow a standard format really, with <workspace_name> and <app_scheme> any arbitrary values I chose as I like - they need not match any other value in any other aspect of the app and its artifacts/resources.
Meaning, the following will always work:

```
XCODE_WORKSPACE: "ios/App/Kuku.xcworkspace"
XCODE_SCHEME: "Kuku"
```

Is my understanding correct?

**Reply by icarusdust** [reply]
* Not entirely, your local project which was prepared on your machine is in Github together with its workspace and scheme/s and Codemagic uses them, does not create anything by itself.
* Yes (M1 machines)
* The main thing here is not the env var but its value, you can name it as BLA_BLA_BLA or anything you want but its values must be correct.  Or you can skip using env vars and use your workspace and scheme directly in scripts. These steps do not have anything to do with Codemagic, these are just simple way of working with env vars or Xcode project.

**Reply by icarusdust** [reply]
If you open your project in VS code, you will see your workspace name if you have pods installed. And if you go to .xcodeproj, you should see your scheme/s. Generally, these are all set in your local Xcode.

**Reply by morsagmon** [reply]
Thanks for the clarification!
Which folders/files are holding this info (so that I can make sure they're not gitignored)?

**Reply by icarusdust** [reply]
They cannot be in .gitignore, a workspace holds your project or projects.

**Reply by icarusdust** [reply]
I strongly suggest you to check the links I sent above, so that you can get familiar with Xcode workspace/schemes and even projects.
